### PR TITLE
[Model] Initial Support for Chameleon

### PR DESCRIPTION
**Pull request by ywang96**
This PR kicks off the effort to add support for Chameleon - Mixed-Modal Early-Fusion Foundation Models from Meta AI. Currently its goal is to match the `transformers` capability to generate text only from text + images.

This PR itself adds `ChameleonForConditionalGeneration` for text-to-text inference. Fully functional vision language inference support with VQVAE will be added in the next PR.

Notable differences between `ChameleonForConditionalGeneration` and `LlamaForCausalLM`
- qk_layernorm
- swin norm 
---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by ywang96** [comment]
This PR is ready for review - it's based on the implementation in https://github.com/huggingface/transformers/pull/31534.

Note: the `transformers` PR hasn't been merged yet - if there's any major update for the language model itself in that PR, I will address it in the next PR if this one gets merged before that.

**Comment by jacobkahn** [comment]
@ywang96 — note the diff in https://github.com/huggingface/transformers/pull/31534#issuecomment-2207354677 will likely be applied before merge, but that no other changes are likely to be made.

**Comment by ywang96** [comment]
> @ywang96 — note the the diff in [huggingface/transformers#31534 (comment)](https://github.com/huggingface/transformers/pull/31534#issuecomment-2207354677) will likely be applied before merge, but that no other changes are likely to be made.

@jacobkahn Thanks for the heads up!

File: vllm/transformers_utils/configs/chameleon.py
Comment (xwjiang2010) [inline vllm/transformers_utils/configs/chameleon.py:4]: curious, why do we copy over some `PretrainedConfig` while leaving out others? Like for `LlavaConfig`, we just used the one from huggingface.

**Comment by xwjiang2010** [comment]
Can we add an example?
Can we add a correctness test to compare with hf result (once it's ready)? We can comment out the compare part for now and just make sure that the code has at least some CI run through.

**Comment by xwjiang2010** [comment]
Should we add the model to the supported vlm model list?

File: vllm/transformers_utils/configs/chameleon.py
Comment (ywang96) [reply vllm/transformers_utils/configs/chameleon.py:4]: I'll change to import it once `transformer` has a new release with the PR merged, otherwise we still need this copied over for now.

**Comment by ywang96** [comment]
@xwjiang2010 Thanks for the quick review!

> Can we add a correctness test to compare with hf result (once it's ready)?

The issue is that we will need to wait for the transformers release to be able to test the hf result, and so far Ive been testing it locally with that branch. I also plan to add a correctness test in the next PR where we add the full text + image support.

> Should we add the model to the supported vlm model list?

Since this PR is only functional in text-to-text, I don't think we should add it to the supported vlm model list yet.

File: vllm/transformers_utils/configs/chameleon.py
Comment (xwjiang2010) [reply vllm/transformers_utils/configs/chameleon.py:4]: Leave a TODO then?

Comment (ywang96) [reply vllm/transformers_utils/configs/chameleon.py:4]: Sounds good

**Comment by ywang96** [comment]
Given there's still ongoing discussion on the original `transformers` PR, I'm going to mark this back to draft until that PR is merged to modify accordingly.

**Comment by AGIGOAT** [comment]
@ywang96 the huggingface PR has been merged

https://github.com/huggingface/transformers/pull/31534#issuecomment-2207354677

**Comment by ywang96** [comment]
This PR has passed model test locally and is ready to be reviewed. cc @xwjiang2010.

File: vllm/model_executor/models/__init__.py
Comment (ywang96) [inline vllm/model_executor/models/__init__.py:18]: This was left unmatched on purpose since HF hasn't fixed the model repo. 

See  https://github.com/huggingface/transformers/issues/32098

Comment (xwjiang2010) [reply vllm/model_executor/models/__init__.py:18]: Add a comment/todo in the file?

**Comment by xwjiang2010** [comment]
Thank you @ywang96.
I did some manual testing. One weird thing that shows up is that 30B in this code doesn't generate anything meaningful in my case (7B works).

Example code:
```
import os
import subprocess

from vllm import LLM

def run():
    # llm = LLM(model="/home/ray/.cache/huggingface/hub/models--facebook--chameleon-7b/snapshots/0f474b71631bcaa20d20d56909066e56cb8a4372/")
    llm = LLM(model="/home/ray/.cache/huggingface/hub/models--facebook--chameleon-30b/snapshots/1c95500025a2a51183c7bb8f410ecd3d84e4cb2c/")

    prompt = "Tell me about tangram."


    outputs = llm.generate({
        "prompt": prompt,
    })

    for o in outputs:
        generated_text = o.outputs[0].text
        print(generated_text)

run()
```

7B gives me: "A tangram is a flat, square puzzle containing seven pieces, each"
30B gives me: "oku<reserved15193> Participationpers earliest noble inhal Wii des meritsassertilet Multi bulanRabrough"

I then switched to huggingface to try out both versions: Both gives me meaningful outputs.

**Comment by ywang96** [comment]
> Thank you @ywang96. I did some manual testing. One weird thing that shows up is that 30B in this code doesn't generate anything meaningful in my case (7B works).
> 
> Example code:
> 
> ```
> import os
> import subprocess
> 
> from vllm import LLM
> 
> def run():
>     # llm = LLM(model="/home/ray/.cache/huggingface/hub/models--facebook--chameleon-7b/snapshots/0f474b71631bcaa20d20d56909066e56cb8a4372/")
>     llm = LLM(model="/home/ray/.cache/huggingface/hub/models--facebook--chameleon-30b/snapshots/1c95500025a2a51183c7bb8f410ecd3d84e4cb2c/")
> 
>     prompt = "Tell me about tangram."
> 
> 
>     outputs = llm.generate({
>         "prompt": prompt,
>     })
> 
>     for o in outputs:
>         generated_text = o.outputs[0].text
>         print(generated_text)
> 
> run()
> ```
> 
> 7B gives me: "A tangram is a flat, square puzzle containing seven pieces, each" 30B gives me: "oku Participationpers earliest noble inhal Wii des meritsassertilet Multi bulanRabrough"
> 
> I then switched to huggingface to try out both versions: Both gives me meaningful outputs.

I think there's something wrong with the swin norm implementation in this PR since it's only enabled for the 30B model. Will look into it!

**Comment by ywang96** [comment]
@xwjiang2010 The swin norm issue is fixed now - I ran your example with `chameleon-30b` with the following code
``` python
from vllm import LLM, SamplingParams
import torch

model_path = "facebook/chameleon-30b"
llm = LLM(model=model_path, dtype=torch.bfloat16)

greedy_params = SamplingParams(temperature=0.0, max_tokens=100)
prompt = "Tell me about tangram."
output = llm.generate(prompt, greedy_params)

print(output[0].outputs[0].text)
```

This gives:
```
Tangram is a puzzle that consists of a square cut into five pieces: a large square, a medium-sized square, a small square, a medium-sized triangle, and a small triangle. The goal of the puzzle is to arrange the pieces into a square shape.

The tangram puzzle has been around for centuries and is believed to have originated in China. It is a popular puzzle that is enjoyed by people of all ages and is often used as a tool for developing
```

**Review by xwjiang2010 — APPROVED** [review]
Awesome! Thanks.

**Review by simon-mo — APPROVED** [review]

**Review by mgoin — APPROVED** [review]
Awesome, excited to give it a try!

**Pull request merged**
